### PR TITLE
[WIP] Rework of Examples Section

### LINKED ISSUE #75: Permission denied error for config directory in examples
I used the examples docker compose file and with some small modifications (domain name set) I get the following errors from the nextcloud page:

`Can't write into config directory!

This can usually be fixed by <a href="https://docs.nextcloud.com/server/11/go.php?to=admin-dir_permissions" target="_blank" rel="noreferrer">giving the webserver write access to the config directory</a>.`

The following errors are shown on the nextcloud-fpm container console:

`PHP message: {"reqId":"J8dbbzhk41yQ6bR4\/K\/+","remoteAddr":"172.18.0.6","app":"PHP","message":"fopen(\/var\/www\/html\/config\/config.php): failed to open stream: No such file or directory at \/var\/www\/html\/lib\/private\/Config.php#230","level":3,"time":"2017-05-18T19:09:06+00:00","method":"GET","url":"\/","user":"--","version":""}
PHP message: {"reqId":"J8dbbzhk41yQ6bR4\/K\/+","remoteAddr":"172.18.0.6","app":"PHP","message":"chmod(): No such file or directory at \/var\/www\/html\/lib\/private\/Config.php#233","level":3,"time":"2017-05-18T19:09:06+00:00","method":"GET","url":"\/","user":"--","version":""}" while reading response header from upstream, client: 172.18.0.6, server: , request: "GET / HTTP/1.1", upstream: "fastcgi://172.18.0.3:9000", host: "[...].ddns.net"
2017/05/18 19:09:06 [error] 5#5: *1 FastCGI sent in stderr: "PHP message: {"reqId":"J8dbbzhk41yQ6bR4\/K\/+","remoteAddr":"172.18.0.6","app":"PHP","message":"chmod(): No such file or directory at \/var\/www\/html\/lib\/private\/Log\/File.php#119","level":3,"time":"2017-05-18T19:09:06+00:00","method":"GET","url":"\/","user":"--","version":""}" while reading upstream, client: 172.18.0.6, server: , request: "GET / HTTP/1.1", upstream: "fastcgi://172.18.0.3:9000", host: "[...].ddns.net"
2017/05/18 19:11:08 [error] 5#5: *56 FastCGI sent in stderr: "PHP message: {"reqId":"tJuu66dEOjjKrdvZuVC3","remoteAddr":"172.18.0.1","app":"PHP","message":"touch(): Unable to create file \/var\/www\/html\/config\/config.php because Permission denied at \/var\/www\/html\/lib\/private\/Config.php#229","level":3,"time":"2017-05-18T19:11:08+00:00","method":"GET","url":"\/","user":"--","version":""}
PHP message: {"reqId":"tJuu66dEOjjKrdvZuVC3","remoteAddr":"172.18.0.1","app":"PHP","message":"fopen(\/var\/www\/html\/config\/config.php): failed to open stream: No such file or directory at \/var\/www\/html\/lib\/private\/Config.php#230","level":3,"time":"2017-05-18T19:11:08+00:00","method":"GET","url":"\/","user":"--","version":""}
PHP message: {"reqId":"tJuu66dEOjjKrdvZuVC3","remoteAddr":"172.18.0.1","app":"PHP","message":"chmod(): No such file or directory at \/var\/www\/html\/lib\/private\/Config.php#233","level":3,"time":"2017-05-18T19:11:08+00:00","method":"GET","url":"\/","user":"--","version":""}" while reading response header from upstream, client: 172.18.0.1, server: , request: "GET / HTTP/1.1", upstream: "fastcgi://172.18.0.3:9000", host: "172.18.0.2"`

This looks like wrong permissions inside the container. (I usually would spend more time with it but I think your docker experience is better than mine and I think others would also profit from this information)

Can you help me on that and extend the readme file or an examples documentation for the things which cannot be found in an existing manual (like setting correct env vars)

### LINKED ISSUE #86: example docker-compose does not work on ubuntu 14.04.5
I am unable to get the docker-compose example working on ubuntu 14.04.5 (which i'm running at my home installed on bare metal) with docker and docker-compose versions:

```
Client:
 Version:      17.05.0-ce
 API version:  1.29
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:06:06 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.05.0-ce
 API version:  1.29 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:06:06 2017
 OS/Arch:      linux/amd64
 Experimental: false

docker-compose version 1.11.2, build dfed245
docker-py version: 2.1.0
CPython version: 2.7.6
OpenSSL version: OpenSSL 1.0.1f 6 Jan 2014
```

I have copied the docker-compose.yml and nginx.conf files to my system, created an environment file to set my domain and mysql passwords (with mysql variables that i added into the docker-compose.yml file).

At this point I set my **DOMAIN** to be the real domain name of my server (i.e. a public domain name).

I executed ``docker-compose up -d``, which failed because the compose file included a docker network that did not exist (**nginx-proxy**).

Next, I created the docker network: ``sudo docker network create nginx-proxy``, which succeeded. Now running the above docker-compose up completed without visible errors.

On attempting to open the web page, I receive HTTP 503 errors. Not working.

I tried changing my domain from the public domain name, to a local domain name (which works only on my private network), docker-compose down and up, and same result - HTTP 503.

Lastly, I changed my domain name to ``localhost`` and this time I am able to get to a Nextcloud page, but instead of a login, I see an error message which says:

```
Can't write into config directory!

This can usually be fixed by <a href="https://docs.nextcloud.com/server/11/go.php?to=admin-dir_permissions" target="_blank" rel="noreferrer">giving the webserver write access to the config directory</a>.
```

I am not sure what else to do at this point, there really isn't any documentation to help me understand what I should be using for a **DOMAIN**, or whether this should work on my version of Ubuntu and Docker.

### LINKED ISSUE #95: nextcloud:fpm Request Entity Too Large
I am running a docker-compose setup using nextcloud:fpm, nginx, nginx-proxy and docker-letsencrypt-nginx-proxy-companion (based on the example in .example, but without a non-sqlite db, redis or collabora). This works for the most part, however I cannot upload files larger than 2MB. According to #32 docker-letsencrypt-nginx-proxy-companion can cause this, but adding a `client_max_body_size` setting to the default or VIRTUAL_HOST-specific vhost.d files has no effect.

Could this be caused by the fast_cgi in the fpm module? I tried adding `fastcgi_param PHP_VALUE "upload_max_filesize = 4000M \n  post_max_size=4000M";` to the server config but this doesn't appear to have any effect either.

### LINKED ISSUE #98: Docker-compose examples uses outdated docker-letsencrypt-nginx-proxy-companion image
https://github.com/nextcloud/docker#https---ssl-encryption links to **JrCs** docker image, but the example docker-compose uses an image created by alastaircoote (image hasn't been update for 6 months).
I'd suggest switching to JrCs' image as he's still keeping it up-to-date.

### LINKED ISSUE #103: Rework of Examples Section
This is a little to do list to improve the examples section:

- [ ] split the examples into subfolders
- [ ] docker-compose example:
  - [ ] use JrCs proxy companion #98 
  - [ ] Use volumes instead of mounted folders 
  - [ ] add proxy settings for upload size #95
  - [ ] don't use an external network as it is not needed when the proxy is defined in the same file. Use multiple networks in the file instead
  - [ ] add example for apache version
  - [ ] change mounts to only `/var/www/html` since it's simpler and less error-prone (would prevent that "config directory not writeable"-error #75 #86) 
  - [ ] remove container names (not needed)
  - [ ] remove `${DOMAIN}`, it isn't explained anywhere
  - [ ] rework the `README.md` to explain the example in detail
  - [x] in fpm version the fpm addition to the tag is missing `nextcloud:fpm` (fixed in #112)
  - [ ] Add example with "indiehosters"-style 2nd container running a cron job #123 
  - [ ] Don't use `volumes_from` in the fpm example since it isn't supported in docker-compose file version 3 
- [ ] derived Dockerfiles:
  - [ ] Add examples for both cases apache and fpm
  - [ ] Add example for preview generation dependencies #63  
  - [ ] Add examples for manipulating the sources #91 #104

### LINKED ISSUE #104: [example docker-compose] Mounting nginx.conf fails with remote docker
I try to use your example docker-compose. I just modified the domain, passwords and email to all use variables. I did put the `nginx.conf` file next to the `docker-compose.yml` file (and checked the permissions of those file).

But when I do
```
DOCKER_HOST=tcp://...:2376 DOCKER_TLS_VERIFY=1 \
DOMAIN="..." PASSWORD="..." EMAIL="..." docker-compose up
```
every container launches correctly except the web container:
```
ERROR: for web  Cannot start service web: oci runtime error: container_linux.go:247: starting container process caused "process_linux.go:359: container init caused \"rootfs_linux.go:54: mounting \\\"/home/didier/Code/docker/nextcloud/nginx.conf\\\" to rootfs \\\"/var/lib/docker/overlay2/770c4e28e5f59e3b7e1ff3fcbacecf011bffcbffda9853a95b412557f6c026f4/merged\\\" at \\\"/var/lib/docker/overlay2/770c4e28e5f59e3b7e1ff3fcbacecf011bffcbffda9853a95b412557f6c026f4/merged/etc/nginx/nginx.conf\\\" caused \\\"not a directory\\\"\""
: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type
ERROR: Encountered errors while bringing up the project.
```

Do you have any idea on how I can fix this ? I found this error a lot on the web but no working fix.

### PR DESCRIPTION
**Pull request by tilosp**
Closes #103 Closes #98 Closes #95 Closes #75 Closes #86 Closes #104

TODO:
- [x] split the examples into subfolders
- [x] docker-compose example:
  - [x] use JrCs proxy companion #98 
  - [x] Use volumes instead of mounted folders 
  - [x] add proxy settings for upload size #95
  - [x] don't use an external network as it is not needed when the proxy is defined in the same file. Use multiple networks in the file instead
  - [x] add example for apache version
  - [x] change mounts to only `/var/www/html` since it's simpler and less error-prone (would prevent that "config directory not writeable"-error #75 #86) 
  - [x] remove container names (not needed)
  - [x] remove `${DOMAIN}`, it isn't explained anywhere
  - [x] Add example with "indiehosters"-style 2nd container running a cron job #123 
  - [x] Don't use `volumes_from` in the fpm example since it isn't supported in docker-compose file version 3
- [x] derived Dockerfiles:
  - [x] Add examples for both cases apache and fpm
  - [x] Add build tests

### CODE REVIEW

**Comment by SnowMB** [comment]
Hey @tilosp,

thanks for starting to work on the examples section. Here a few thoughts on the folder structure.

### current version
```
examples\apache\dockerfiles\Dockerfile.cron
examples\apache\dockerfiles\supervisord.conf
examples\fpm\dockerfiles\Dockerfile.cron
examples\fpm\dockerfiles\supervisord.conf
```
This has the disadvantage that additional files would clutter the structure and users might ask themself what are these files for (e.g. when s.o. wants to build an `Dockerfile.imap` and there is the `supervisord.conf` file). It could get worse when more examples are added.

I would prefer to differentiate between the base images (fpm/apache) in the last subfolder. That leaves three options:

### option 1
```
examples\dockerfiles\cron\Dockerfile.apache
examples\dockerfiles\cron\Dockerfile.fpm
examples\dockerfiles\cron\supervisord.conf
```

This has the advantage, that the additional files would not be redundant. As long as the additional files can be used in all base images it would be a very clean folder structure.

### option 2
```
examples\dockerfiles\cron\apache\Dockerfile
examples\dockerfiles\cron\apache\supervisord.conf
examples\dockerfiles\cron\fpm\Dockerfile
examples\dockerfiles\cron\fpm\supervisord.conf
```
Reintroduces redundency, but in cases where different additional files for the base images are needed, their would be no conflicts. Also it is possible to run the default commands `docker build .` or `docker-compose up -d` directly in the folder.

### option 3
```
examples\dockerfiles\cron\apache\Dockerfile
examples\dockerfiles\cron\fpm\Dockerfile
examples\dockerfiles\cron\supervisord.conf
```
This last option could be used to avoid the redundency issue by moving matching files to the parent folder. But at the same time the Dockerfiles get dependant on these directories.

### My opinion:
__option 1__ is useful, if someone dislikes using to many subfolders 😉
Personally I prefer __option 2__ because it seems to be the cleanest one.

**Comment by tilosp** [comment]
@SnowMB i agree with you that **option 2** is the cleanest one and i am fine with using many sub-directories.
And by the way `supervisord.conf` is not redundant because one of them starts fpm and the other one apache.

**Comment by SnowMB** [comment]
You're right. Missed that. But you unterstood what I tried to explain 😁

File: .examples/docker-compose/with-nginx-proxy/apache/caching.php [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?
+$CONFIG = array (
+  'memcache.local' => '\\OC\\Memcache\\APCu',
```
Comment (tilosp) [inline .examples/docker-compose/with-nginx-proxy/apache/caching.php]: @SnowMB this line is redundant, it is already part of the image [apcu.config.php](https://github.com/nextcloud/docker/blob/master/.config/apcu.config.php)

Comment (SnowMB) [inline .examples/docker-compose/with-nginx-proxy/apache/caching.php]: done

**Comment by SnowMB** [comment]
Hey @tilosp,

I got a quick question:

You have an app folder for each example creating a derived image copying the '*.php' files. Wouldn't it be easier mounting them in the 'docker-compose.yml' file and using the default image?

**Comment by tilosp** [comment]
@SnowMB 
There is one advantage.
Mounting them is just fine when the docker engine is on the same machine as the docker client.
But if you use a docker engine on a remote machine mounting will fail because the files are only on the client side so that the remote docker engine can't mount them. This problem came up in #104.
When building an image the client transfers the build context (all file which are not excluded by a `.dockerignore` file) to the docker engine to build it there.

I forgot to do the same for the `nginx.conf` which i will do later.

**Comment by SnowMB** [comment]
Ah, I see. Great solution for that case! 👍

**Comment by SnowMB** [comment]
@tilosp I reworked the proxy examples to fit your style and added a readme file. Please review and maybe add a few words to the readme.

**Comment by tilosp** [comment]
@SnowMB please take a look.

I removed the example with collabora for now because it does not work. The problem is that the nginx proxy is not setup to handle WebSocket connection and collabora requires WebSocket connection. I would like to merge this PR without a collabora example, we can add one in a later PR.

But we should add a link to the main readme which links to the example readme. Maybe you have an idea where exactly this link should go.

I would also suggest a rework of [README.md#running-this-image-with-docker-compose](https://github.com/nextcloud/docker/blob/master/README.md#running-this-image-with-docker-compose). Maybe by just replacing it with a link to the docker-compose examples

**Comment by tilosp** [comment]
ping @SnowMB

File: .examples/README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+## docker-compose
+In `docker-compose` additional services are bundled to create a complete nextcloud installation. The examples are designed to run out-of-the-box.
+Before running the examples you have to modify the `db.env` and `docker-compose.yml` file and fill in your custom information.
+
+The docker-compose examples make heavily use of dereived Dockerfiles to add configuration files into the containers. This way they should also work on remote docker systems as _Docker for Windows_. When running docker-compose on the same host as the docker daemon, another possibility would be to simply mount the files in the volumes section in the `docker-compose.yml` file.
+
+
+### insecure
+
+To use this example complete the following steps:
```
Comment (SnowMB) [inline .examples/README.md]: add a sentence why this setup is insecure and that it should not be used in production.

Comment (tilosp) [reply .examples/README.md]: done

**Pull request merged**